### PR TITLE
feat(#862): PartOps Slice C — explode parts into scene nodes + join back

### LINKED ISSUE #862: PartOps Slice C — Explode segmented parts into separate scene nodes and join back
## Goal

Add Blender-style explode/join workflow for segmented or submesh-based parts.

Parent epic: #859

## Scope

- Add `Explode Parts` from either existing submeshes or segmentation groups.
- Create one child scene node/entity per part.
- Preserve world transform, material, and vertex attributes.
- Add configurable explode offset for visual inspection.
- Add `Join Parts` to merge selected exploded entities into one mesh, baking transforms into vertices.
- Reuse/extend #626 join semantics for vertex declaration and material conflict handling.

## Acceptance Criteria

- [ ] Exploded parts are independent selectable scene nodes.
- [ ] Moving exploded parts and joining back bakes transforms correctly.
- [ ] Undo restores pre-explode scene state.
- [ ] Joined mesh exports as one entity with expected submeshes/materials.

## Tests

- Explode a multi-part fixture.
- Move one part, join, assert baked vertex positions.
- Undo/redo scene state test.

### PR DESCRIPTION
**Pull request by fernandotonon**
## Summary

PartOps epic **#859**, **Slice C (#862)** — Blender-style explode/join for segmented parts, built on the Slices A+B split (already merged in #923).

- **Explode** a multi-submesh mesh (typically one produced by "Split into Parts") into independent, selectable/movable scene nodes, pushed outward by a configurable distance.
- **Join** a multi-selection of part nodes back into one fused mesh, baking each node's world transform into the vertices (same-material submeshes coalesce).
- Both fully undoable (Ctrl+Z).

## What's in it

| Piece | Role |
|---|---|
| `PartOpsScene.{h,cpp}` | Scene-level Ogre adapter above `PartOpsMesh`. `explodeEntity()` → one single-submesh `Ogre::Mesh` per part (attributes/material/skeleton+bone-assignments + part name preserved) + outward `SubMeshOps::explodeOffsets`. `joinEntities()` → bakes each node's `_getFullTransform()` into positions (inverse-transpose into normals/tangents), merges via `SubMeshOps::joinParts`. Pure builders — no scene mutation. |
| `ExplodePartsCommand` | Undoable: redo destroys the fused node, creates N sibling part nodes at `srcTransform + local-frame offset`, reselects them; undo restores the fused node bound to the resident original mesh. |
| `JoinPartsCommand` | Undoable: redo captures each part's mesh + node TRS, destroys the parts, creates one fused node at the origin (positions already world-baked); undo recreates every part. Both clear the SelectionSet before destroying entities (SplitMeshCommand dangling-sub-entity-ref rationale). |
| `PartOpsController` | `explodeSelected(distance)` / `joinSelected()` + `canExplode` (one multi-submesh selection) / `canJoin` (2+ selected) gate props. |
| `qml/PropertiesPanel.qml` | Object-mode Inspector **"Explode / Join Parts"** section: logarithmic explode-distance slider (0–10, default 0.1) + Explode/Join buttons. |

Breadcrumbs `mesh.parts.explode` / `mesh.parts.join`.

## Design decisions (confirmed with the maintainer)

- **Explode consumes the source** — the fused entity is replaced by N sibling nodes; undo restores the single fused mesh. Matches the split epic's swap-the-node pattern and makes Join the obvious inverse.
- **Join operates on the current multi-selection** — works for any set of part nodes, not just ones from a single explode.
- Skeletons are **not** reconciled on join (documented limitation) — the merged mesh is static geometry.

## Tests

- `SubMeshOps_test.cpp`: new `JoinBakesRotationIntoPositionsAndNormals` (proves positions **and** normals rotate under a 90° transform — the "bakes transforms correctly" AC) + single-part passthrough.
- `ExplodePartsCommand_test.cpp` / `JoinPartsCommand_test.cpp`: no-Ogre error-branch coverage (ctor/text, initial state, unresolvable-entity redo, undo-before-redo no-op), mirroring `SplitMeshCommand_test.cpp`. The full GL round-trip stays with the Ogre-gated CLI tests (arriving with Slice E).
- All 24 pass locally; verified end-to-end in the GUI (explode → move → join → undo).

## Scope

Explode/join **CLI + MCP parity is Slice E (#864)**; the print-peg dialog is Slice D (#863) — both remain open per the epic.

Closes #862.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added Object-mode “Explode / Join Parts” tools with an adjustable explode distance, selection-based enablement, and completion feedback.
  * Introduced undo/redo for explode and join, preserving transforms and reselecting created parts.
* **Bug Fixes**
  * Improved joined-mesh results for mirrored transforms by fixing triangle winding and tangent handedness.
  * Enhanced submesh material handling and submesh reading accuracy.
* **Tests**
  * Added unit tests for explode/join commands plus additional transform/material edge cases.
* **Documentation**
  * Expanded PartOps guidance with scene-level explode/join semantics and command behavior.

### CODE REVIEW

File: src/commands/ExplodePartsCommand.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    // rationale: sub-entity refs would dangle through getResolvedEntities()).
+    if (auto* sel = SelectionSet::getSingleton())
+        sel->clearList();
+
+    // Destroy the fused source node (frees its name so undo can recreate it).
+    Ogre::Entity* src = resolveSourceEntity();
+    if (src) {
+        Ogre::SceneNode* node = src->getParentSceneNode();
+        if (node)
+            mgr->destroySceneNode(node, /*destroyChildrenFirst=*/true);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/ExplodePartsCommand.cpp]: Fixed in d8d5e76. Explode now captures the source node's parent and reparents the part nodes back under it (and the restored fused node on undo), via `Manager::reparentNode` + an explicit local-TRS restore. Nodes that have child nodes are rejected up front with a clear error, since the command doesn't serialise arbitrary subtrees.

File: src/commands/JoinPartsCommand.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    // Drop selection refs before destroying source entities (SplitMeshCommand
+    // rationale: dangling sub-entity refs crash the next selection query).
+    if (auto* sel = SelectionSet::getSingleton())
+        sel->clearList();
+
+    // Destroy the source part nodes (frees their names for undo to recreate).
+    for (const auto& sp : mSources) {
+        if (Ogre::SceneNode* node = mgr->getSceneNode(QString::fromStdString(sp.name)))
+            mgr->destroySceneNode(node, /*destroyChildrenFirst=*/true);
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/JoinPartsCommand.cpp]: Fixed in d8d5e76. Join now records each part's parent group + local TRS and reparents the restored parts back under it on undo; parts with child nodes are rejected up front. Grouped parts round-trip through join/undo.

File: src/PartOpsScene.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+    std::vector<SubMeshOps::JoinPart> parts;
+    parts.reserve(entities.size());
+    for (Ogre::Entity* e : entities) {
+        if (!e || !e->getMesh()) {
+            out.error = QStringLiteral("a selected object is not a mesh");
+            return out;
+        }
+        SubMeshOps::JoinPart jp;
+        if (!PartOpsMesh::readSubMeshes(e, jp.subMeshes)) {
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/PartOpsScene.cpp:116]: Fixed in d8d5e76. `PartOpsMesh::readSubMeshes` now overrides each editable submesh's material with the entity's effective per-SubEntity material (`SubEntity::getMaterialName`), so a Material-Mode assignment survives split/explode/join and join no longer coalesces visibly-distinct parts.

File: src/PartOpsScene.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                            .arg(QString::fromStdString(e->getName()));
+            return out;
+        }
+        // Bake the part's FULL world transform (node hierarchy included) so a
+        // moved/rotated exploded part lands in the right place in the merged
+        // mesh. _getFullTransform is the node's world affine; joinParts uses
+        // it for positions and its inverse-transpose for normals/tangents.
+        Ogre::SceneNode* node = e->getParentSceneNode();
+        jp.transform = node ? Ogre::Matrix4(node->_getFullTransform())
+                            : Ogre::Matrix4::IDENTITY;
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/PartOpsScene.cpp:127]: Fixed in d8d5e76. `SubMeshOps::joinParts` detects a negative-determinant (orientation-reversing) transform and reverses triangle winding + flips tangent handedness so a mirrored/negative-scaled part joins front-facing. Covered by JoinReversesWindingUnderMirrorTransform / JoinKeepsWindingUnderNonMirrorTransform.

File: src/commands/ExplodePartsCommand.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    // Reselect the created part nodes so the user can immediately move/join.
+    if (auto* sel = SelectionSet::getSingleton()) {
+        sel->clearList();
+        for (const auto& n : mPartNodeNames) {
+            if (Ogre::SceneNode* pn = mgr->getSceneNode(QString::fromStdString(n)))
+                sel->append(pn);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/ExplodePartsCommand.cpp]: Fixed in 864d2cd. Both commands now CREATE-THEN-DESTROY in every redo()/undo(): the replacement node(s) are built and entity-created (and validated) before the old node is destroyed — their names never collide with the node being replaced, so they coexist momentarily. A creation failure rolls back the new nodes and leaves the original intact, so the scene is never orphaned. Same pattern applied to JoinPartsCommand redo/undo.

File: src/commands/ExplodePartsCommand.h [hunk 6]
```diff
@@ -0,0 +1,80 @@
+#ifndef EXPLODE_PARTS_COMMAND_H
+#define EXPLODE_PARTS_COMMAND_H
+
+#include <QUndoCommand>
+#include <QString>
+
+#include <OgreMesh.h>
+#include <OgreVector3.h>
+
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/ExplodePartsCommand.h:10]: Fixed in 864d2cd — added #include <OgreQuaternion.h> to ExplodePartsCommand.h.

File: src/PartOpsController.cpp [hunk 7]
```diff
@@ -42,6 +45,24 @@ bool PartOpsController::hasSelection() const
...
+        return false;
+    // Nothing to explode from a single-submesh mesh.
+    return entities.first()->getMesh()->getNumSubMeshes() >= 2;
+}
+
+bool PartOpsController::canJoin() const
+{
+    const auto* sel = SelectionSet::getSingleton();
+    return sel && sel->getResolvedEntities().size() >= 2;
+}
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/PartOpsController.cpp:74]: Fixed in 864d2cd. canJoin() and joinSelected() now count only real mesh entities (skip null / non-mesh resolved entries), so a 1-mesh+1-null selection no longer enables Join or builds a single-part command. The QML button reads canJoin directly, so its enabled state is corrected too.

File: src/commands/ExplodePartsCommand.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/commands/ExplodePartsCommand.cpp]: Fixed in bce1f96. JoinPartsCommand::undo() is now all-or-nothing: every restored part node/entity is validated, a partial failure rolls back the parts already created, and the fused node is destroyed only after ALL parts are back — so a failed undo can no longer replace the fused result with a subset of its parts.

File: src/PartOpsMesh.cpp [hunk 8]
```diff
@@ -18,7 +19,26 @@ bool PartOpsMesh::readSubMeshes(Ogre::Entity* entity,
...
+    const unsigned short n = entity->getNumSubEntities();
+    for (unsigned short i = 0; i < n && i < outSubMeshes.size(); ++i) {
+        Ogre::SubEntity* se = entity->getSubEntity(i);
+        if (!se)
+            continue;
+        const std::string effective = se->getMaterialName();
+        if (!effective.empty())
+            outSubMeshes[i].materialName = effective;
+    }
+    return true;
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/PartOpsMesh.cpp:41]: Added in bce1f96 — PartOpsMesh_material_coverage_test.cpp (GL-gated). A 2-submesh entity gets a runtime SubEntity material override on submesh 0; the test asserts readSubMeshes() reports the effective override for sub0 and the base material for the un-overridden sub1, verifying the positional mapping. (Materials must exist first, else Ogre resolves an unknown name to BaseWhite — the test creates them up front.)

File: src/PartOpsMesh_material_coverage_test.cpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+void clearSceneNodes()
+{
+    if (!Manager::getSingletonPtr())
+        return;
+    auto nodes = Manager::getSingleton()->getSceneNodes();
+    for (auto* node : nodes) {
+        Manager::getSingleton()->destroyAllAttachedMovableObjects(node);
+        Manager::getSingleton()->destroySceneNode(node);
+    }
```
[the remark this answers is not in the document]
Comment (fernandotonon) [reply src/PartOpsMesh_material_coverage_test.cpp]: Fixed in 2143ea1. The fixture now uses unique per-test names for the node/mesh/materials and TearDown destroys only those (the node + its attached objects, the mesh, the 3 materials) — no global scene sweep, no fixed-name removal.
